### PR TITLE
P2P handshake handling

### PR DESCRIPTION
**Pull request by prestonvanloon**
The basic idea of this PR is to implement the following:

On new peer connection, send a "handshake" message to the peer. The peer should also respond with a "handshake" message. If the two peers do not agree on some of the information in the handshake, they should disconnect and not try to reconnect with each other.

Banning peers is blocked by https://github.com/libp2p/go-libp2p/issues/274

### CODE REVIEW

File: shared/p2p/negotiation.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+type negotiator struct {
+	contractAddress string
+}
+
+// Notifee ...
+func (n *negotiator) Notifee() inet.Notifiee {
+	return &inet.NotifyBundle{
+		ConnectedF: func(net inet.Network, conn inet.Conn) {
+			go func() {
```
Comment (prestonvanloon) [inline shared/p2p/negotiation.go]: Without this goroutine, this method is blocking forever while waiting for a response on that protocol, but the peer's protocol handler is never hit.

With this goroutine, it seems the stream is immediately reset and/or some junk data is manifested from somewhere. The error is that the protobuf cannot be unmarshalled

**Comment by raulk** [comment]
Hey @prestonvanloon, this usage of the libp2p API is not correct, that's why you're observing weird effects. The job of the connection manager is to keep the connection count in check to avoid killing commercial routers.

What you want to do is attach your notify directly to the host's network:

```
host.Network().Notify(...)
```

You can:
1. pass in a custom struct that implements the `inet.Notifiee` interface, or
2. use the convenience struct `NotifyBundle`; it wrongly resides in go-libp2p-connmgr right now, but is being hoisted to go-libp2p-core soon as part of https://github.com/libp2p/go-libp2p/issues/602.

Once you do that, here's the key: you want to open the stream **via the `host` itself** (`host.NewStream(peer, protocol)`), not through the `conn`. The conn doesn't know about multistream negotiation, and will open a raw stream for you but will not negotiate the protocol. The call to `SetProtocol(string)` won't kick off that negotiation either, that's actually an internal API that has, unfortunately, leaked to the user API. So the result is that you have a raw stream with no protocol pinned on it on the wire, therefore the other end doesn't notify the stream handler for that protocol.

I'm sorry our APIs are a bit messy here. I'm taking this issue as an "experience report" and once we settle the "core" refactor, we'll embark on a mission to make our APIs more cohesive and intuitive.

File: shared/p2p/negotiation.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+)
+
+const handshakeProtocol = prysmProtocolPrefix + "/handshake"
+
+func setupPeerNegotiation(h host.Host, contractAddress string) {
+	h.Network().Notify(&inet.NotifyBundle{
+		ConnectedF: func(net inet.Network, conn inet.Conn) {
+			log.Debug("Checking connection to peer")
+
+			s, err := h.NewStream(context.Background(), conn.RemotePeer(), handshakeProtocol)
```
Comment (prestonvanloon) [inline shared/p2p/negotiation.go]: @raulk, this is blocking / never opens the stream. Am I missing something?

**Comment by prestonvanloon** [comment]
Thanks for the feedback @raulk. I'm still unable to create a stream between the peers. I'm seeing that host.NewStream is blocking indefinitely on both sides.

File: shared/p2p/negotiation.go (hunk 2 above)
Comment (raulk) [reply shared/p2p/negotiation.go]: Could you try to create the stream in a goroutine? "never block the callback" is the problem I think you're seeing here.

Comment (prestonvanloon) [reply shared/p2p/negotiation.go]: That works, thanks!

File: shared/p2p/handshake_handler.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+func setHandshakeHandler(host host.Host, contractAddress string) {
+	host.SetStreamHandler(handshakeProtocol, func(stream inet.Stream) {
+		defer stream.Close()
+		log.Debug("Handling handshake stream")
+		w := ggio.NewDelimitedWriter(stream)
+		defer w.Close()
+
+		hs := &pb.Handshake{DepositContractAddress: contractAddress}
+		if err := w.WriteMsg(hs); err != nil {
+
```
Comment (terencechain) [inline shared/p2p/handshake_handler.go]:
```suggestion
```

**Pull request merged**